sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionService.java

```diff
@@ -0,0 +1,147 @@
+        if (!clusterService.localNode().isMasterNode()) {
+            return;
+        }
+        clusterService.submitStateUpdateTask(request.cause,
+            new AckedClusterStateUpdateTask<ClusterStateUpdateResponse>(request, listener) {
+
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    MetaData metaData = currentState.metaData();
+                    MetaData.Builder mdBuilder = MetaData.builder(metaData);
+
+                    UserDefinedFunctionsMetaData functions = putFunction(
+                        metaData.custom(UserDefinedFunctionsMetaData.TYPE),
+                        request.metaData,
+                        request.replace
+                    );
+                    mdBuilder.putCustom(UserDefinedFunctionsMetaData.TYPE, functions);
+                    return ClusterState.builder(currentState).metaData(mdBuilder).build();
+                }
+
+                @Override
+                protected ClusterStateUpdateResponse newResponse(boolean acknowledged) {
+                    return new ClusterStateUpdateResponse(acknowledged);
+                }
+            });
+    }
+
+    @VisibleForTesting
+    static UserDefinedFunctionsMetaData putFunction(@Nullable UserDefinedFunctionsMetaData oldMetaData,
+                                                    UserDefinedFunctionMetaData functionMetaData,
+                                                    boolean replace) {
+        if (oldMetaData == null) {
+            return UserDefinedFunctionsMetaData.of(functionMetaData);
+        } else {
+            if (!replace && oldMetaData.contains(functionMetaData)) {
+                throw new UserDefinedFunctionAlreadyExistsException(functionMetaData);
+            }
+            // create a new instance of the metadata, to guarantee the cluster changed action.
+            UserDefinedFunctionsMetaData newMetaData = UserDefinedFunctionsMetaData.newInstance(oldMetaData);
+            newMetaData.put(functionMetaData);
+            return newMetaData;
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        UserDefinedFunctionsMetaData oldMetaData = event.previousState().getMetaData()
+            .custom(UserDefinedFunctionsMetaData.TYPE);
+        UserDefinedFunctionsMetaData newMetaData = event.state().getMetaData()
+            .custom(UserDefinedFunctionsMetaData.TYPE);
+
+        // Check if user defined functions got changed
+        if ((oldMetaData == null && newMetaData == null) || (oldMetaData != null && oldMetaData.equals(newMetaData))) {
+            return;
+        }
+
+        Map<String, Map<FunctionIdent, FunctionImplementation>> schemaFunctions = newMetaData.functionsMetaData().stream()
+            .map(UserDefinedFunctionFactory::of)
+            .collect(groupingBy(f -> f.info().ident().schema(), toMap(f -> f.info().ident(), f -> f)));
```

Comment:
couldn't the loop be changed to go through the list instead? I think the map creation is not really necessary